### PR TITLE
[SPARK-33438][SQL] Eagerly init objects with defined SQL Confs for command `set -v`

### PR DESCRIPTION
**Pull request by linhongliu-db**
### What changes were proposed in this pull request?
In Spark, `set -v` is defined as "Queries all properties that are defined in the SQLConf of the sparkSession".
But there are other external modules that also define properties and register them to SQLConf. In this case,
it can't be displayed by `set -v` until the conf object is initiated (i.e. calling the object at least once).

In this PR, I propose to eagerly initiate all the objects registered to SQLConf, so that `set -v` will always output
the completed properties.

### Why are the changes needed?
Improve the `set -v` command to produces completed and  deterministic results

### Does this PR introduce _any_ user-facing change?
`set -v` command will dump more configs

### How was this patch tested?
existing tests

### CODE REVIEW

**Comment by linhongliu-db** [comment]
cc @cloud-fan

**Review by srowen — COMMENTED** [review]
How important is this? would it work 'later' in a job after these are used and initialized?
It just adds a non-trivial amount of complexity and not sure of the implications of eagerly loading.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -3811,6 +3811,8 @@ class SQLConf extends Serializable with Logging {
    * definition contains key, defaultValue and doc.
    */
   def getAllDefinedConfs: Seq[(String, String, String, String)] = sqlConfEntries.synchronized {
+    // Eagerly load conf objects to force the conf entries registering
+    SparkConfRegisterLoader.load(Utils.getContextOrSparkClassLoader)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3815]: shall we make sure this is called globally once?

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala [hunk 2]
```diff
@@ -106,6 +114,16 @@ class SQLConfSuite extends QueryTest with SharedSparkSession {
       Row("spark.sql.groupByOrdinal", "false"))
   }
 
+  test("eager init") {
+    checkAnswer(
+      sql("SET").where(s"key = '${FakeConf.FAKE_CONFIG.key}'").select("key", "value"),
+      Nil)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala:120]: why this query has no result?

**Comment by cloud-fan** [comment]
Seems like a reasonable bug fix of the `SET -v` command, but the bug should be rare. cc @viirya @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (linhongliu-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3815]: sure, will do

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala (hunk 2 above)
Comment (linhongliu-db) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala:120]:
```
    // (In Hive, "SET" returns all changed properties while "SET -v" returns all properties.)
    // Queries all key-value pairs that are set in the SQLConf of the sparkSession.
```
according to the comments in `SetCommand`, `SET` should not output the config that not changed.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala:120]: ah i see

**Comment by linhongliu-db** [comment]
@srowen, Thanks for reviewing.
I think this PR is a kind of bug fix since the `set -v` doesn't output expected results in some cases. Also, even though it would work later after the SQLConf objects are used, it still may mislead users if the outputs of `set -v` are different when run multiple times.
Regarding the complexity, actually, this PR almost has no side effect on existing code. even though there are exceptions during  eagerly loading, it will be catched and won't break the spark

File: core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import java.net.URL
+
+import scala.collection.JavaConverters._
+import scala.io.Source
+import scala.reflect.runtime.{universe => ru}
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+
+object SparkConfRegisterLoader extends Logging {
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: Why do we need this class for initing configs eagerly? We cannot simply load them as `PythonSQLUtils` does?
https://github.com/apache/spark/blob/d99d0d27be875bba692bcfe376f90c930e170380/sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala#L48-L58

Comment (linhongliu-db) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: @maropu, awesome! I was not aware of this in `PythonSQLUtils`. But I'm still worried about the maintenance if other external modules also have registered SQLConf. And this code assumes the hive module is already included which is not suitable for all other moudles.

Comment (maropu) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: Yea, we might be able to brush up it based on the current implementation. But, IMO it'd be better to avoid loading classes from a resource file.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: Yea it's easier to just list all the known config objects, but I agree with @linhongliu-db that some modules are optional and we should only load the objects from modules when the modules are loaded. FYI all modules under the `external` directory are not loaded by default, and if they start to define configs, we will need this framework. It's also true that other spark vendors may define their own configs.

This is similar to the data source registry resource files, so I don't have a big concern here as this is not the first time.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: We should let `PythonSQLUtils` use this new framework though.

**Review by viirya — COMMENTED** [review]
honestly this looks complex, especially the issue it tries to solve seems rare. And it seems loading arbitrary object and we are not sure about what will be loaded actually.

File: core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala (hunk 3 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: This looks an overkill. So do we plan to separate the options for `external` modules? I would just simply do `ScalaReflection.mirror.staticModule("org.apache.spark.sql.hive.HiveUtils")` in `SetCommand`.

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: This allows people to define SQL configs outside of Spark, e.g. spark packages, vendor-specific configs. We should probably make it clear in the PR title and description. If we just want to fix the bug, we can hardcode the loading of `HiveUtis` class like @HyukjinKwon mentioned.

How about we separate this PR into 2:
1. One PR to add `SparkConfRegisterLoader`, which can be used in SetCommand and `PythonSQLUtils`. This fixes the bug for SetCommand
2. One PR to add the resource file loading, which allows people to define SQL configs outside of Spark. We can discuss more if this feature is reasonable in that PR.

Comment (linhongliu-db) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: @cloud-fan sure, I can do this

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: > One PR to add the resource file loading, which allows people to define SQL configs outside of Spark.

We should probably make `SQLConf.buildConf` and `SQLConf.buildStaticConf` as proper developer APIs first

Comment (maropu) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: > One PR to add the resource file loading, which allows people to define SQL configs outside of Spark.

Ah, I see. Anyway, I also think It is better to separate the bug fix and the feature improvement as @cloud-fan said above.

Comment (viirya) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: Fixing `SetCommand` first sounds good to me.

Comment (linhongliu-db) [reply core/src/main/scala/org/apache/spark/util/SparkConfRegisterLoader.scala:29]: thanks all for reviewing, I'll update this PR to focus on bug fixing only. and open another to discuss resource file loading.

**Review by cloud-fan — APPROVED** [review]

**Comment by linhongliu-db** [comment]
cc @viirya @maropu @HyukjinKwon, this PR is updated based on discussion

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -3799,6 +3801,27 @@ class SQLConf extends Serializable with Logging {
...
+      definedConfsLoaded = true
+      // Force to register static SQL configurations
+      StaticSQLConf
+      try {
+        // Force to register SQL configurations from Hive module
+        val symbol = ScalaReflection.mirror.staticModule("org.apache.spark.sql.hive.HiveUtils")
+        ScalaReflection.mirror.reflectModule(symbol).instance
+      } catch {
+        case NonFatal(e) =>
+          logWarning("Cannot load sql configurations from hive module", e)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3820]: nit: This logging reads like there is a problem. But if we don't build with Hive, it sounds no problem at all? Maybe rephrase to "SQL configurations from Hive module is not loaded."?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -3799,6 +3801,27 @@ class SQLConf extends Serializable with Logging {
     }
   }
 
+  private var definedConfsLoaded = false
+  /**
+   * init [[StaticSQLConf]] and [[org.apache.spark.sql.hive.HiveUtils]] so that all the defined
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3806]: nit: init -> Init

**Review by viirya — APPROVED** [review]
Minor comments, otherwise looks okay.

**Review by viirya — COMMENTED** [review]
> How was this patch tested?
newly added and existing tests

BTW, I don't see any new test was added. Need to update the description?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.1, and branch-3.0.

**Comment by maropu** [comment]
late lgtm
